OpenRA.Game/Graphics/SpriteFont.cs

```diff
@@ -65,48 +69,83 @@ void PrecacheColor(Color c, string name)
+			location += new float2(0, size);
+
+			// Calculate positions in screen pixel coordinates
+			var screenContrast = (int)(contrastOffset * deviceScale);
+			var screen = new int2((int)(location.X * deviceScale + 0.5f), (int)(location.Y * deviceScale + 0.5f));
+			var contrastVector = new float2(screenContrast, screenContrast);
+			foreach (var s in text)
+			{
+				if (s == '\n')
+				{
+					location += new float2(0, size);
+					screen = new int2((int)(location.X * deviceScale + 0.5f), (int)(location.Y * deviceScale + 0.5f));
+					continue;
+				}
+
+				var g = glyphs[Pair.New(s, Color.Black)];
+
+				// Convert screen coordinates back to UI coordinates for drawing
+				if (g.Sprite != null)
+				{
+					var contrastSprite = contrastGlyphs[Tuple.Create(s, contrastColor, screenContrast)];
+					Game.Renderer.RgbaSpriteRenderer.DrawSprite(contrastSprite,
+						(screen + g.Offset - contrastVector) / deviceScale,
+						contrastSprite.Size / deviceScale);
+				}
+
+				screen += new int2((int)(g.Advance + 0.5f), 0);
+			}
+		}
+
 		public void DrawText(string text, float2 location, Color c)
 		{
 			// Offset from the baseline position to the top-left of the glyph for rendering
 			location += new float2(0, size);
 
-			var p = location;
+			// Calculate positions in screen pixel coordinates
+			var screen = new int2((int)(location.X * deviceScale + 0.5f), (int)(location.Y * deviceScale + 0.5f));
 			foreach (var s in text)
 			{
 				if (s == '\n')
 				{
 					location += new float2(0, size);
-					p = location;
+					screen = new int2((int)(location.X * deviceScale + 0.5f), (int)(location.Y * deviceScale + 0.5f));
 					continue;
 				}
 
 				var g = glyphs[Pair.New(s, c)];
+
+				// Convert screen coordinates back to UI coordinates for drawing
 				if (g.Sprite != null)
 					Game.Renderer.RgbaSpriteRenderer.DrawSprite(g.Sprite,
-						new float2(
-							(int)Math.Round(p.X * deviceScale + g.Offset.X, 0) / deviceScale,
-							p.Y + g.Offset.Y / deviceScale),
-						g.Sprite.Size / deviceScale);
+					(screen + g.Offset).ToFloat2() / deviceScale,
+					g.Sprite.Size / deviceScale);
 
-				p += new float2(g.Advance / deviceScale, 0);
+				screen += new int2((int)(g.Advance + 0.5f), 0);
 			}
 		}
 
-		float3 Rotate(float3 v, float sina, float cosa, float2 offset)
+		float2 Rotate(float2 v, float sina, float cosa, float2 offset)
 		{
-			return new float3(
+			return new float2(
 				v.X * cosa - v.Y * sina + offset.X,
-				v.X * sina + v.Y * cosa + offset.Y,
-				0);
+				v.X * sina + v.Y * cosa + offset.Y);
 		}
 
 		public void DrawText(string text, float2 location, Color c, float angle)
 		{
 			// Offset from the baseline position to the top-left of the glyph for rendering
+			// All positions are calculated in UI coordinates
 			var offset = new float2(0, size);
 			var cosa = (float)Math.Cos(-angle);
 			var sina = (float)Math.Sin(-angle);
 
+			// TODO: round positions to screen pixels as above
```

Comment:
This TODO has already been resolved. I will remove it when addressing any other review comments.